### PR TITLE
feat(server): DEBUG MEMORYFUZZ command

### PR DESCRIPTION
**Pull request by dranikpg**
WIP
PUSHED TO CHECK OUT FROM AWS

### CODE REVIEW

File: src/server/debugcmd.h [hunk 1]
```diff
@@ -19,13 +22,35 @@ class DebugCmd {
...
+ private:
+  struct KeyRange {
+    uint64_t from, to;
+    std::string_view prefix;
+  };
+
+  using PopulateFunction = std::function<void(std::string_view prefix, const PopulateBatch& batch)>;
+
+  // Run populate function on given range in batches
+  void RunPopulate(KeyRange range, DebugCmd::PopulateFunction func);
```
Comment (adiholden) [inline src/server/debugcmd.h:51]: DebugCmd:: before PopulateFunction can be removed

File: src/server/debugcmd.h [hunk 2]
```diff
@@ -19,13 +22,35 @@ class DebugCmd {
...
   void Populate(CmdArgList args);
-  void PopulateRangeFiber(uint64_t from, uint64_t len, std::string_view prefix, unsigned value_len);
+  void MemoryFuzz(CmdArgList args);
   void Reload(CmdArgList args);
   void Replica(CmdArgList args);
   void Load(std::string_view filename);
   void Inspect(std::string_view key);
   void Watched();
 
+ public:
```
Comment (adiholden) [inline src/server/debugcmd.h:32]: Is there a reason why you added another section of public and private in this class and not add to the existing ?

Comment (dranikpg) [reply src/server/debugcmd.h:32]: That's still just experimental code, not ready for any polishment 😅

**Comment by dranikpg** [comment]
@romange 
I'm simulating a slowly growing heap, but I can't reproduce above 20% over-use with this method. Any ideas? 

Currently, I can get to about 16-17% in just a few seconds with 20GB on 12 threads. Anything approaching 20 always flattens out.

**Comment by dranikpg** [comment]
In any case, if we want just an integration test for observing defragmentation under load, we can use a shrinking heap

https://github.com/dragonflydb/dragonfly/pull/509/files#r1030738038

**Comment by romange** [comment]
Me and @adiholden discussed just yesterday that we need a similar method (extend debug populate to generate random uncompressable data) but in the context of snapshot compression. 

Now, in the context of memory fragmentation, I did not deep dive into your code, but i do not think that random methods are needed there. I can suggest a precise method on how to explore the weaknesses of slab allocators:

```cpp
for (i in range(1, 50)) {
  auto binlen = mimalloc_bin_size(i);    //  you can easily discover first 50 binsizes by checking out mi_good_size
  auto ptrs = allocate_fuckload_blobs(N, binlen);   // N order of 64K-256K to create megabytes of allocated data
  deallocate_uniformly(0.9, ptrs);   // this should leave all (most) mimalloc blocks committed but "used" will decrease by 90%
}
```

I also think it will be quicker to test inside dragonfly tests and not with the external tests. Finally, there are mimalloc options that need to be enabled to prohibit delayed decommits (see my discussions with daan).

**Pull request closed without merging**